### PR TITLE
bpo-37807: Make hash() return a non-negative number.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue37807

### CODE REVIEW

File: Doc/library/functions.rst [hunk 1]
```diff
@@ -705,6 +705,10 @@ are always available.  They are listed here in alphabetical order.
       truncates the return value based on the bit width of the host machine.
       See :meth:`__hash__` for details.
 
+   .. versionchanged:: 3.9
+      :fucn:`hash` returns a non-negative number.
```
Comment (sir-sigurd) [inline Doc/library/functions.rst]:
```suggestion
      :func:`hash` returns a non-negative number.
```

File: Objects/typeobject.c [hunk 2]
```diff
@@ -6492,7 +6492,12 @@ slot_tp_hash(PyObject *self)
        Py_hash_t.  Therefore our transformation must preserve values that
        already lie within this range, to ensure that if x.__hash__() returns
        hash(y) then hash(x) == hash(y). */
-    h = PyLong_AsSsize_t(res);
+    if (Py_SIZE(res) < 0) {
+        h = PyLong_AsSsize_t(res);
+    }
+    else {
+        h = (Py_ssize_t)PyLong_AsSize_t(res);
```
Comment (sir-sigurd) [inline Objects/typeobject.c:6500]: It looks like signed overflow can happen here.

**Comment by mdickinson** [comment]
Please could you also update the documentation for the numeric hash? The old docs are no longer true with this change.

File: Objects/typeobject.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c:6500]: Yes, but it is everywhere in evaluating and using hashes.

**Comment by serhiy-storchaka** [comment]
I do not know what words to use for documenting the changes in the algorithm. I have added a comment, so now [Doc/library/stdtypes.rst](https://github.com/serhiy-storchaka/cpython/edit/unsigned-hash/Doc/library/stdtypes.rst?pr=%2Fpython%2Fcpython%2Fpull%2F15199) can be edited via browser. @mdickinson, could you please help with documentation?

But I am not sure that this change should be done. I do not know what code will need `hash()` returning a non-negative number.

**Comment by rhettinger** [comment]
I withdrew this suggestion due to the suggestion that the implementation details became guaranteed.

**Pull request closed without merging**